server/src/main/java/io/crate/types/DataTypes.java

```diff
@@ -538,33 +538,47 @@ public static void register(int id, Writeable.Reader<DataType<?>> dataType) {
     }
 
     /**
-     * Compares any two {@link DataType} by their IDs.
+     * Compares any two {@link DataType} by their IDs and names. For ObjectTypes, their inner types are compared only if the names are the same.
+     * The rest of the DataTypes are compared by their IDs.
      * The parameters of the data types, if they have any, are ignored.
      */
-    public static boolean isSameType(DataType<?> left, DataType<?> right) {
+    public static boolean isCompatibleType(DataType<?> left, DataType<?> right) {
         if (left.id() != right.id()) {
             return false;
         } else if (isArray(left)) {
-            return isSameType(
+            return isCompatibleType(
                 ((ArrayType<?>) left).innerType(),
                 ((ArrayType<?>) right).innerType());
-        } else {
-            return true;
+        } else if (left.id() == ObjectType.ID) {
+            var l = (ObjectType) left;
+            var r = (ObjectType) right;
+            for (var lName : l.innerTypes().keySet()) {
+                var lInner = l.innerTypes().get(lName);
```

Comment:
I think iterating over the entrySet would be cheaper as it avoids the key lookup on the `get` call

Reply:
Thanks! fixed